base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -2651,6 +2710,57 @@ fn convert_to_option_bounds<T: RangeBounds<u64>>(bounds: T) -> (Option<u64>, Opt
     (start, end)
 }
 
+/// Process a batch of blocks for PayRef migration
+fn process_payref_for_height<B: BlockchainBackend>(
+    db: Arc<RwLock<B>>,
+    height: u64,
+    metadata_at_start: ChainMetadata,
+    finalize: bool,
+) -> Result<PayrefRebuildStatus, ChainStorageError> {
+    debug!(target: LOG_TARGET, "[PayRef] Processing index rebuilding for height {}", height);
+
+    // Get all outputs for this block
+    let read_txn = db
+        .read()
+        .map_err(|_e| ChainStorageError::AccessError("Read lock on blockchain backend failed".into()))?;
+
+    let binding = read_txn.fetch_chain_header_by_height(height)?;
+    let header = binding.header();
+    let block_hash = header.hash();
+    let outputs = read_txn.fetch_outputs_in_block(&block_hash)?;
+    drop(read_txn);
+
+    let mut write_txn = db
+        .write()
+        .map_err(|_e| ChainStorageError::AccessError("Write lock on blockchain backend failed".into()))?;
+
+    let mut txn = DbTransaction::new();
+    for output in &outputs {
```

Comment:
what happens if the block was reorged while this is happening?
This entire fn needs a single write lock